### PR TITLE
drivers: ps2: microchip: Low power and wakeup enabled

### PR DESCRIPTION
**Pull request by Manimaran-A**
ps2 driver updated to support low power and wakeup feature

### CODE REVIEW

File: drivers/ps2/ps2_mchp_xec.c [hunk 1]
```diff
@@ -42,11 +48,40 @@ struct ps2_xec_config {
...
 };
 
+#ifdef CONFIG_PM_DEVICE
+static void ps2_xec_pm_policy_state_lock_get(struct ps2_xec_data *data,
+					       enum ps2_pm_policy_state_flag flag)
+{
+	if (atomic_test_and_set_bit(data->pm_policy_state_flag, flag) == 0) {
+		pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	}
+}
```
[the remark this answers is not in the document]
Comment (Manimaran-A) [reply drivers/ps2/ps2_mchp_xec.c:74]: Both two (ps2_xec_pm_policy_state_lock_get and ps2_xec_pm_policy_state_lock_put) APIs are used in more than one place in the file hence we added wrapper function for better readability.

**Comment by jvasanth1** [comment]
@albertofloyd can you help to review?

**Review by jvasanth1 — DISMISSED** [review]

File: drivers/ps2/ps2_mchp_xec.c [hunk 2]
```diff
@@ -205,14 +253,82 @@ static int ps2_xec_enable_interface(const struct device *dev)
...
+					&devcfg->wakerx_gpio,
+					GPIO_INT_MODE_EDGE | GPIO_INT_TRIG_LOW);
+				if (ret < 0) {
+					LOG_ERR("Fail to enable PS2 wake interrupt(ret %d)", ret);
+					return ret;
+				}
+			}
+		} else {
+			regs->CTRL &= ~MCHP_PS2_CTRL_EN;
+#ifdef CONFIG_PINCTRL
```
Comment (albertofloyd) [inline drivers/ps2/ps2_mchp_xec.c:313]: Now that pinmux is gone do we still these kind of checks?

**Review by albertofloyd — DISMISSED** [review]

File: drivers/ps2/ps2_mchp_xec.c (hunk 2 above)
Comment (Manimaran-A) [reply drivers/ps2/ps2_mchp_xec.c:313]: CONFIG_PINCTRL is still a configurable option in dts hence keeping it.

**Review by jvasanth1 — DISMISSED** [review]

File: drivers/ps2/ps2_mchp_xec.c (hunk 2 above)
Comment (gmarull) [reply drivers/ps2/ps2_mchp_xec.c:313]: This is no longer the case, please drop as the platform only works with pinctrl now. Driver must select pinctrl.

**Review by gmarull — CHANGES_REQUESTED** [review]
Drop conditional pinctrl

**Review by gmarull — CHANGES_REQUESTED** [review]
Drop conditional pinctrl

File: drivers/ps2/ps2_mchp_xec.c (hunk 2 above)
Comment (Manimaran-A) [reply drivers/ps2/ps2_mchp_xec.c]: Ok, CONFIG_PINCTRL has been removed.

File: drivers/ps2/ps2_mchp_xec.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/ps2/ps2_mchp_xec.c:75]: still not justified why we need this.

[the remark this answers is not in the document]
Comment (Manimaran-A) [reply drivers/ps2/ps2_mchp_xec.c]: Removed the wrapper and used policy API

**Review by albertofloyd — DISMISSED** [review]

**Review by jvasanth1 — DISMISSED** [review]

File: drivers/ps2/ps2_mchp_xec.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/ps2/ps2_mchp_xec.c]: this is not solved yet, you just removed the wrapper, but **why** is atomic needed at all here?

**Review by gmarull — DISMISSED** [review]
issue not fixed yet

File: drivers/ps2/ps2_mchp_xec.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (Manimaran-A) [reply drivers/ps2/ps2_mchp_xec.c]: Updated the code by removing atomic check.

File: drivers/ps2/ps2_mchp_xec.c [hunk 3]
```diff
@@ -32,17 +37,35 @@ struct ps2_xec_config {
...
+	PS2_PM_POLICY_FLAG_COUNT,
+};
+#endif
+
 struct ps2_xec_data {
 	ps2_callback_t callback_isr;
 	struct k_sem tx_lock;
+#ifdef CONFIG_PM_DEVICE
+	ATOMIC_DEFINE(pm_policy_state_flag, PS2_PM_POLICY_FLAG_COUNT);
+#endif
```
Comment (gmarull) [inline drivers/ps2/ps2_mchp_xec.c:65]:
```suggestion
```

File: drivers/ps2/ps2_mchp_xec.c [hunk 4]
```diff
@@ -32,17 +37,35 @@ struct ps2_xec_config {
...
 };
 
+#ifdef CONFIG_PM_DEVICE
+enum ps2_pm_policy_state_flag {
+	PS2_PM_POLICY_TX_FLAG,
+	PS2_PM_POLICY_RX_FLAG,
+	PS2_PM_POLICY_FLAG_COUNT,
+};
+#endif
+
```
Comment (gmarull) [inline drivers/ps2/ps2_mchp_xec.c:59]:
```suggestion
```

Comment (Manimaran-A) [reply drivers/ps2/ps2_mchp_xec.c]: Unused enum removed from the source.

File: drivers/ps2/ps2_mchp_xec.c (hunk 3 above)
Comment (Manimaran-A) [reply drivers/ps2/ps2_mchp_xec.c]: Deleted the ATOMIC_DEFINE.

**Comment by Manimaran-A** [comment]
@gmarull @jvasanth1 @albertofloyd  Could you please help to review and approve the PR

**Review by jvasanth1 — APPROVED** [review]

**Review by albertofloyd — APPROVED** [review]

**Pull request merged**